### PR TITLE
MinGW: PRINTF_FORMAT_ARG workaround

### PR DESCRIPTION
**Pull request by yadij**
MinGW produces warnings about not supporting the printf '%zu' format
code when the 'printf' format archetype is used to validate code.

Checking against the 'gnu_printf' format archetype avoids this bad
warning. We accept the lower rate of error detection since other OS
builds verify against the 'printf' format archetype.

We also removed undocumented, inconsistent, and presumably unused
support for providing custom PRINTF_FORMAT_ARG macros during Squid
build. This removal simplifies code.

### CODE REVIEW

File: compat/compat_shared.h [hunk 1]
```diff
@@ -248,7 +248,17 @@ const char * squid_strnstr(const char *s, const char *find, size_t slen);
...
-#else /* !__GNU__ */
+#elif __GNUC__
+#if !defined(PRINTF_FORMAT_ARG1)
+#define PRINTF_FORMAT_ARG1 __attribute__ ((format (gnu_printf, 1, 2)))
+#endif
+#if !defined(PRINTF_FORMAT_ARG2)
+#define PRINTF_FORMAT_ARG2 __attribute__ ((format (gnu_printf, 2, 3)))
+#endif
+#if !defined(PRINTF_FORMAT_ARG3)
+#define PRINTF_FORMAT_ARG3 __attribute__ ((format (gnu_printf, 3, 4)))
```
Comment (rousskov) [inline compat/compat_shared.h:259]: > [GCC](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-format-function-attribute): The parameter archetype determines how the format string is interpreted, and should be printf, ... or gnu_printf. On MinGW targets, ms_printf [is] also present.
>
> * archetype values such as printf refer to the formats accepted by the system’s C runtime library, 
> * values prefixed with ‘gnu_’ always refer to the formats accepted by the GNU C Library.
> * On Microsoft Windows targets, values prefixed with ‘ms_’ refer to the formats accepted by the msvcrt.dll library

> PR description: Some build profiles of GCC require the gnu_printf format attribute

If I am interpreting the implications of this PR description statement correctly, some GCC builds warn about some printf() %code used by the official code. That means that %code is not "accepted by the system’s C runtime library". If we enable gnu_printf, we will silence the warning because the offending %code is "accepted by the GNU C Library". However, that warning is correct if the build is _not_ actually using "the GNU C Library". In those cases, we should not silence the warning. Instead, we should change build parameters and/or Squid code.

To avoid such false positives, we can use gnu_printf _only_ when the built is using "the GNU C Library". The proposed `!__clang__ && __GNUC__` condition is different -- as this PR illustrates, `__GNUC__` does not even mean "built using GCC". Can we adjust that condition to match "built using GCC and the GNU C Library"?

Also, _if_ this change is triggered by Windows builds that use msvcrt.dll, then it sounds like we should enable `ms_printf` (for those GCC builds) instead.

In summary, we should use the format attribute that matches the actual build (rather than the format attribute that silences warnings): If a GCC build uses msvcrt.dll, we should use `ms_printf` attribute. If a GCC build uses GNU C Library, we should use `gnu_printf` attribute. Can we adjust these defines to match the current build?

Comment (yadij) [reply compat/compat_shared.h:259]: That GCC documentation does not tell the whole picture. The archetypes have changed over time as what was "standard" changed. The exact profile for "printf" depends both on which ```-std=``` option is specified, and on macros which can be explicitly set.

Unfortunately MinGW sets a macro to downgrades the "printf" support to the ANSI compliant C89 standard set. Not a C++ compliant set of %codes. The "gnu_printf" archetype is always a superset of the ```-std=``` support, so we can safely use that in its place to get ```%zu``` etc from C++11 and later.

We cannot use the "ms_printf" archetype because that only defines the minimal ANSI set along with a bunch of weird ```%I32...``` and ```%I64...``` prefixed codes for MSVC we should not be polluting Squid with.

Comment (rousskov) [reply compat/compat_shared.h:259]: > MinGW sets a macro to downgrades the "printf" support to the ANSI compliant C89 standard set

I assume the above statement is about what is deemed supported by the "printf" archetype, not what is supported by the actual printf() implementation in that specific MinGW environment.

> The "gnu_printf" archetype is always a superset of the `-std=` support, so we can safely use that in its place to get `%zu` etc from C++11 and later.

AFAICT, these archetypes do not control what is supported by printf(). They control what the compiler warns about when printf() is called.

Given that assumption, I do not understand the "we can safely use a larger set" logic: We are not trying to find the archetype that includes our %codes. We are trying to find the archetype that matches what is actually _supported_ by the current build environment! Obviously, larger archetype sets will produce fewer warnings, but our goal is not fewer warnings. Our goal is detecting unsupported %codes (i.e. correct warnings). Fewer (wrong) warnings is just a side effect.

It is, of course, possible that there is just no archetype that matches our needs and we have to use gnu_printf, even though we risk missing silencing some correct warnings.

> Unfortunately MinGW sets a macro to downgrades the "printf" support

Can we make `gnu_printf` use specific to MinGW then, to reduce the impact of this problem on other `__GNUC__` builds where this "downgrade" of `printf` archetype does not happen?

Comment (yadij) [reply compat/compat_shared.h]: Done in [9762652](https://github.com/squid-cache/squid/pull/1390/commits/97626527cffaa9343c139f37ab7e1e8417fbadea) now that we have the ```os/mingw.h``` header to place the hack.

File: compat/os/mingw.h [hunk 2]
```diff
@@ -32,5 +32,17 @@
 #include <windows.h>
 #endif
 
+#if defined(__GNUC__)
+#if !defined(PRINTF_FORMAT_ARG1)
+#define PRINTF_FORMAT_ARG1 __attribute__ ((format (gnu_printf, 1, 2)))
```
[the remark this answers is not in the document]
Comment (yadij) [reply compat/os/mingw.h]: Done in [0fb27a5](https://github.com/squid-cache/squid/pull/1390/commits/0fb27a5de85daada875e5321c3cfdaa4cece3a5a)

**Review by rousskov — COMMENTED** [review]
I still do not insist on any changes. And please feel free to merge without even commenting on any of change requests.

File: compat/compat_shared.h [hunk 3]
```diff
@@ -239,15 +239,15 @@ const char * squid_strnstr(const char *s, const char *find, size_t slen);
 #endif
 
 #if __GNUC__
-#if !defined(PRINTF_FORMAT_ARG1)
```
[the remark this answers is not in the document]
Comment (yadij) [reply compat/compat_shared.h:242]: Please feel free to add a statement to the description. I am not sure how to word it other than one of those "foo is foo" statements you dislike.

[the remark this answers is not in the document]
Comment (yadij) [reply compat/compat_shared.h:242]: I disagree on the need for the supposed change. So will not be adding a TODO here,

[the remark this answers is not in the document]
Comment (rousskov) [reply compat/compat_shared.h:242]: OK, I will try to guess that reason and document it.

File: compat/compat_shared.h [hunk 4]
```diff
@@ -239,15 +239,15 @@ const char * squid_strnstr(const char *s, const char *find, size_t slen);
...
+#if _SQUID_MINGW_
```
[the remark this answers is not in the document]
Comment (rousskov) [reply compat/compat_shared.h:242]: I disagree that the suggested move requires testing with MSVC:

* This PR is already changing code in ways that may directly affect MSVC builds. If we assume, for the sake of the argument, that "we must test with MSVC any changes that might affect MSVC builds" assertion is valid, then we must test this PR with MSVC as well.

* Moving that code clearly improves Squid code quality (even if we make a mistake with the exact MSVC definitions!). Unknown MSVC needs are less important than known Squid code improvements.

* Moving that code is likely to reduce MSVC build problems by reducing the chances that the same macro will be _redefined_, triggering MSVC compiler [warnings](https://godbolt.org/z/beWxbvz88). The current definitions in two files _contradict_ each other AFAICT (although I did not check all the surrounding details/exclusions):

```C++
#if __GNUC__
#if _SQUID_MINGW_
...
#else
#define PRINTF_FORMAT_ARG1 __attribute__ ((format (printf, 1, 2)))
```

```C++
/* gcc doesn't recognize the Windows native 64 bit formatting tags causing
 * the compile fail, so we must disable the check on native Windows.
 */
#if __GNUC__
#define PRINTF_FORMAT_ARG1
```

File: compat/compat_shared.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply compat/compat_shared.h:242]: Done.

File: compat/compat_shared.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (yadij) [reply compat/compat_shared.h:242]: Very well, to make progress done in [13f3770](https://github.com/squid-cache/squid/pull/1390/commits/13f37708202789fc6d827b1adc671dcc27c2a9e3).
